### PR TITLE
Functionality for GAIL

### PR DESCRIPTION
**Pull request by kantneel**
Features
- recording trajectories 
- training GAIL agents with those trajectories

### CODE REVIEW

File: src/modelfree/common/utils.py [hunk 1]
```diff
@@ -43,6 +46,9 @@ def save(self, save_path):
     def load(self):
         raise NotImplementedError()
 
+    def _get_pretrain_placeholders(self):
```
Comment (AdamGleave) [inline src/modelfree/common/utils.py]: Why do we need this? Not part of the standard `BaseRLModel` API AFAIK.

File: src/modelfree/common/utils.py [hunk 2]
```diff
@@ -149,39 +155,114 @@ def _reset_video_recorder(self, metadata=None):
         )
 
 
+class ReporterOutputFormat(KVWriter):
```
Comment (AdamGleave) [inline src/modelfree/common/utils.py]: Why is this here? It's also in `src/modelfree/multi/train_worker.py`, which is the only place I can see it being used. (Would be reasonable to factor out -- although probably not into `util` -- if it was used in multiple places.)

File: src/modelfree/common/utils.py [hunk 3]
```diff
@@ -149,39 +155,114 @@ def _reset_video_recorder(self, metadata=None):
...
+
 def make_session(graph=None):
     tf_config = tf.ConfigProto()
     tf_config.gpu_options.allow_growth = True
     sess = tf.Session(graph=graph, config=tf_config)
     return sess
 
 
-def simulate(venv, policies, render=False):
+class TrajectoryRecorder(object):
```
Comment (AdamGleave) [inline src/modelfree/common/utils.py]: Needs a docstring

File: src/modelfree/common/utils.py [hunk 4]
```diff
@@ -149,39 +155,114 @@ def _reset_video_recorder(self, metadata=None):
...
+    def __init__(self, num_envs, num_policies, num_traj_to_save, traj_dir):
+        self.num_envs = num_envs
+        self.num_policies = num_policies
+        self.num_traj_to_save = num_traj_to_save
+        self.traj_dir = traj_dir
+        if not os.path.isdir(self.traj_dir):
```
Comment (AdamGleave) [inline src/modelfree/common/utils.py]: Better to "to ask for forgiveness than permission" in Python for complex operations like file I/O. Although it probably doesn't matter in this particular case, in general there can be tricky race conditions if you check for file existence in operation and then do something that depends on it in a second operation. Better to just do `os.makedirs(self.traj_dir, exist_ok=True)`.

Comment (AdamGleave) [reply src/modelfree/common/utils.py]: See https://stackoverflow.com/questions/12265451/ask-forgiveness-not-permission-explain for general coding style

File: src/modelfree/common/utils.py (hunk 3 above)
Comment (AdamGleave) [reply src/modelfree/common/utils.py]: Consider making `TrajectoryRecorder` a `MultiVecWrapper`. Then it can be easily dropped into a variety of situations, not just `simulate`.

File: src/modelfree/common/utils.py [hunk 5]
```diff
@@ -149,39 +155,114 @@ def _reset_video_recorder(self, metadata=None):
...
+        data_vals = (rewards, actions, prev_observations)
+        iter_space = itertools.product(enumerate(self.traj_dicts), range(self.num_envs))
+        # iterate over both agents over all environments in VecEnv
+        for (agent_idx, agent_dicts), env_idx in iter_space:
+            for key, val in zip(data_keys, data_vals):
+                agent_dicts[env_idx][key].append(val[agent_idx][env_idx])
+            if dones[env_idx]:
+                ep_len = len(agent_dicts[env_idx]['rewards'])
+                ep_starts = [True] + [False] * (ep_len - 1)
+                self.full_traj_dicts[agent_idx]['episode_starts'].append(np.array(ep_starts))
```
Comment (AdamGleave) [inline src/modelfree/common/utils.py]: Why are we concatenating all the trajectories into one big list, and then having a separate boolean mask for episode starts, rather than having a list of lists (one for each episode)? Maybe GAIL expects it in this format, but we can always convert into this format when loading into GAIL; let's use a sane data format for our project at least.

File: src/modelfree/common/utils.py [hunk 6]
```diff
@@ -149,39 +155,114 @@ def _reset_video_recorder(self, metadata=None):
...
+
+                ep_ret = sum(agent_dicts[env_idx]['rewards'])
+                self.full_traj_dicts[agent_idx]['episode_returns'].append(np.array([ep_ret]))
```
Comment (AdamGleave) [inline src/modelfree/common/utils.py]: 'rewards' seems clearer than 'episode_returns' (I expect this is more GAIL compatibility?)

File: src/modelfree/common/utils.py [hunk 7]
```diff
@@ -149,39 +155,114 @@ def _reset_video_recorder(self, metadata=None):
...
+                    self.full_traj_dicts[agent_idx][key].append(episode_key_data)
+                agent_dicts[env_idx] = defaultdict(list)
+                # so as not to double-count number of episodes
+                self.num_completed += int(agent_idx == 0)
+                if self.num_completed >= self.num_traj_to_save:
+                    self.save_traj(agent_idx)
+
+    def save_traj(self, agent_idx):
+        if self.already_saved[agent_idx]:
+            print(f'Warning: already saved trajectories for agent {agent_idx}')
```
Comment (AdamGleave) [inline src/modelfree/common/utils.py]: Use `logger.warning` instead (if we don't already have a logger, write `logger = logging.getLogger('modelfree.common.utils')` at top of the file)

File: src/modelfree/common/utils.py [hunk 8]
```diff
@@ -149,39 +155,114 @@ def _reset_video_recorder(self, metadata=None):
...
+
+
+def simulate(venv, policies, render=False, record_traj=False, traj_dir='data/',
+             num_traj_to_save=None):
     """
     Run Environment env with the agents in agents
     :param venv(VecEnv): vector environment.
     :param policies(list<BaseModel>): a policy per agent.
     :param render: true if the run should be rendered to the screen
+    :param record_traj: whether to save trajectory data in stable_baselines.GAIL format
```
Comment (AdamGleave) [inline src/modelfree/common/utils.py]: If we make the `TrajectoryRecorder` a wrapper, then we can move this out of `simulate` into the constructor params.

File: src/modelfree/common/utils.py [hunk 9]
```diff
@@ -149,39 +155,114 @@ def _reset_video_recorder(self, metadata=None):
...
+    :param traj_dir: where to save the trajectories
+    :param num_traj_to_save: when to save trajectories since this function is a generator
```
Comment (AdamGleave) [inline src/modelfree/common/utils.py]: Why is this needed? We could just e.g. call `save` when we're done simulating. (I think then it's OK to remove the "already saved" check.) There are plenty of cases where we might not ahead of time know how many trajectories we're going to generate (e.g. do a fixed number of timesteps not episodes, fixed number of wall-clock time, etc.)

File: src/modelfree/common/utils.py [hunk 10]
```diff
@@ -149,39 +155,114 @@ def _reset_video_recorder(self, metadata=None):
...
+        for idx, (policy, obs, state) in enumerate(zip(policies, observations, states)):
             act, new_state = policy.predict(obs, state=state, mask=dones)
             actions.append(act)
             new_states.append(new_state)
         actions = tuple(actions)
         states = new_states
 
         observations, rewards, dones, infos = venv.step(actions)
+        if record_traj:
+            recorder.record_traj(rewards, actions, dones, prev_observations)
```
Comment (AdamGleave) [inline src/modelfree/common/utils.py]: Nitpick: I'd find `prev_observations, actions, rewards, dones` a more intuitive order. We see `prev_observations` first; we choose `actions`; then we receive `rewards` and `dones`.

File: src/modelfree/score_agent.py [hunk 11]
```diff
@@ -24,7 +24,8 @@ def announce_winner(sim_stream):
                 yield game_outcome(info)
 
 
-def get_empirical_score(_run, env, agents, episodes, render=False):
+def get_empirical_score(_run, env, agents, episodes, render=False,
+                        record_traj=False, traj_dir='data/'):
```
Comment (AdamGleave) [inline src/modelfree/score_agent.py]: Yuck! Record trajectory definitely has no job here. This will also be unnecessary if we make it a wrapper.

File: src/modelfree/score_agent.py [hunk 12]
```diff
@@ -57,8 +59,10 @@ def default_score_config():
     agent_a_path = '1'     # path or other unique identifier
     agent_b_type = 'zoo'   # type supported by policy_loader.py
     agent_b_path = '2'     # path or other unique identifier
+    record_traj = False    # record trajectories in stable_baselines.GAIL format
+    traj_dir = 'data/'     # directory to save trajectories to
```
Comment (AdamGleave) [inline src/modelfree/score_agent.py]: Maybe make it a subdirectory by default? e.g. `data/experts`

File: src/modelfree/train.py [hunk 13]
```diff
@@ -142,18 +142,26 @@ def callback(locals, globals):
     model.learn(total_timesteps=total_timesteps, log_interval=1, seed=_seed, callback=callback)
     final_path = osp.join(out_dir, 'final_model')
     _save(model, final_path, save_callbacks)
-    model.sess.close()
+    if isinstance(model, GAIL):
```
Comment (AdamGleave) [inline src/modelfree/train.py]: Can you instead make a PR for Stable Baselines to make the GAIL class expose `sess` from `trpo.sess`? Stable Baselines should have a consistent interface here.

File: src/modelfree/train.py [hunk 14]
```diff
@@ -173,6 +181,22 @@ def sac(batch_size, learning_rate, **kwargs):
                    batch_size=batch_size, learning_rate=learning_rate, **kwargs)
 
 
+@train_ex.capture
+def gail(batch_size, expert_dataset_path, **kwargs):
+    import matplotlib
+    matplotlib.use('pdf')  # ExpertDataset needs this and we don't have tkinter
```
Comment (AdamGleave) [inline src/modelfree/train.py]: Can you also in the PR for GAIL get rid of the `matplotlib` dependency for `ExpertDataset`? It's just used in the `plot` function, which TBH would probably be fine to delete, but if we want to keep it then easy to make that function import matplotlib only when needed. Then we can remove this hack.

File: src/modelfree/train.py [hunk 15]
```diff
@@ -173,6 +181,22 @@ def sac(batch_size, learning_rate, **kwargs):
...
+    from stable_baselines.gail.dataset.dataset import ExpertDataset
+
+    num_proc = _get_mpi_num_proc()
+    if expert_dataset_path is None:
+        raise ValueError("must set expert_dataset_path to use GAIL.")
+    expert_dataset = ExpertDataset(expert_dataset_path)
+    del kwargs['learning_rate']
```
Comment (AdamGleave) [inline src/modelfree/train.py]: Umm... shouldn't this be used somewhere? (Maybe under a different name?)

File: tests/modelfree/test_experiments.py [hunk 16]
```diff
@@ -53,6 +56,13 @@ def test_score_agent(config):
     outcomes = [run.result[k] for k in ['ties', 'win0', 'win1']]
     assert sum(outcomes) == run.config['episodes']
 
+    if 'record_traj' in config:
```
Comment (AdamGleave) [inline tests/modelfree/test_experiments.py]: Change to `if config.get('record_traj', False)`

We don't want this to apply if `'record_traj': False` is in the config.

File: tests/modelfree/test_experiments.py [hunk 17]
```diff
@@ -28,6 +28,7 @@ def test_experiment(experiment):
 SCORE_AGENT_CONFIGS = [
     {'agent_b_type': 'zoo', 'agent_b_path': '2', 'videos': True, 'episodes': 2},
     {'env_name': 'multicomp/KickAndDefend-v0', 'episodes': 1},
+    {'record_traj': True, 'traj_dir': 'test_dir'}
```
Comment (AdamGleave) [inline tests/modelfree/test_experiments.py]: :+1: for test case

File: tests/modelfree/test_experiments.py [hunk 18]
```diff
@@ -86,10 +96,16 @@ def test_score_agent(config):
     {
         'env_name': 'multicomp/SumoHumansAutoContact-v0',
         'adv_noise_params': {'noise_val': 0.1},
-    }
+    },
+    {
+        'rl_algo': 'gail',
+        'num_env': 1,
+        'expert_dataset_path': 'tests/modelfree/SumoAnts_traj/agent_0.npz',
```
Comment (AdamGleave) [inline tests/modelfree/test_experiments.py]: Best to avoid large binary files in repos as it slows down cloning. It's worth it for a test case, but can we maybe make the dataset file smaller? 3 MB is quite a lot (would expect to correspond to many episodes). Also if we only use the `agent_0` one, can we remove the `agent_1` file?

File: src/modelfree/common/utils.py (hunk 1 above)
Comment (kantneel) [reply src/modelfree/common/utils.py]: It is in the `gail-test` branch which has the other gail-related fixes I've been using

File: tests/modelfree/test_experiments.py (hunk 18 above)
Comment (kantneel) [reply tests/modelfree/test_experiments.py]: each file corresponds to only 20 episodes actually. Recall that a file with only 5000 episodes was taking > 1 GB of data in S3. Nonetheless i'll make it like 5 episodes instead lol

Comment (AdamGleave) [reply tests/modelfree/test_experiments.py]: :+1: for a test even just two episodes is enough. (1 maybe not, since there could be edge cases where it fails with >1 episode.)

File: src/modelfree/common/utils.py (hunk 1 above)
Comment (AdamGleave) [reply src/modelfree/common/utils.py]: OK, fair enough, I see it got added in one of the recent PRs for Stable Baselines

File: src/modelfree/common/utils.py [hunk 19]
```diff
@@ -156,6 +161,87 @@ def make_session(graph=None):
...
+                # data_vals always have data for all agents (use agent_idx not dict_idx)
+                agent_dicts[env_idx][key].append(val[agent_idx][env_idx])
+
+            for key in transparency_keys:
+                # infos also always have data for all agents
+                if key not in infos[env_idx][agent_idx]:
+                    continue
+                agent_dicts[env_idx][key].append(infos[env_idx][agent_idx][key])
+            if dones[env_idx]:
+                if self.use_gail_format:
```
Comment (AdamGleave) [inline src/modelfree/common/utils.py]: GAIL compatibility code does not belong in TrajectoryRecorder, makes it harder to understand. Also we should really standardize on a single format for trajectory recording in our code, and do conversions as needed just before feeding into external libraries.

**Comment by AdamGleave** [comment]
@kantneel: the tests were failing because the Stable Baselines commit you switched to missed out many of the hotfixes this repo depends on, e.g. some of my PRs that have not yet been merged into Stable Baselines. In the future please be careful to test things thoroughly when making version changes. I've merged these and some more recent changes now into https://github.com/HumanCompatibleAI/baselines/tree/adv-policies, and have made changes to this PR to fix compatibility issues.

Other than that the code looked fairly good. The main issue was the trajectory recorder was very closely entangled with GAIL. This made sense originally when we were just using it for GAIL, but is unfortunate now it is finding more wide use. So I refactored this. There were also some minor bugs and cosmetic details that I fixed; please see my changes + the review comments above.

There are still a few comments that need to be addressed (mostly docstrings), please go through and ping me when done and I'll take another look.

@decodyng: would appreciate your review on this, especially the changes I've introduced.

File: src/modelfree/train.py [hunk 20]
```diff
@@ -101,6 +102,7 @@ def old_ppo2(_seed, env, out_dir, total_timesteps, num_env, policy,
 def _stable(cls, our_type, callback_key, callback_mul, _seed, env, env_name, out_dir,
             total_timesteps, policy, load_policy, rl_args, victim_index, debug, logger,
             log_callbacks, save_callbacks, log_interval, checkpoint_interval, **kwargs):
+    kwargs = {k: v for k, v in kwargs.items() if k not in rl_args}
```
[the remark this answers is not in the document]
Comment (kantneel) [reply src/modelfree/train.py]: I agree. No, I think I meant it as a way to prevent an error from accidentally happening on the next line, but I'm going to take it out.

**Comment by decodyng** [comment]
@AdamGleave Sure, happy to take a look

File: src/modelfree/score_agent.py [hunk 21]
```diff
@@ -53,23 +53,28 @@ def get_empirical_score(_run, env, agents, episodes, render=False):
...
-    num_env = 1            # number of environments to run in parallel
-    episodes = 20           # number of episodes to evaluate
-    render = True          # display on screen (warning: slow)
-    videos = False         # generate videos
-    video_dir = 'videos/'  # video directory
+    agent_a_type = 'zoo'                # type supported by policy_loader.py
+    agent_a_path = '1'                  # path or other unique identifier
+    agent_b_type = 'zoo'                # type supported by policy_loader.py
+    agent_b_path = '2'                  # path or other unique identifier
+    record_traj = False                 # whether to record trajectories
```
Comment (decodyng) [inline src/modelfree/score_agent.py]: Apropos of the fact that we currently have our default config not using the trajectory recording functionality, do we have or, if we don't have, do we want a test that start up a TrajectoryRecording env, confirm that a JSON is saved properly, that is the expected set of keys, etc? My default would be to think that's a good thing to add, unless there's some general belief to not test things that require running trajectories because that'd be too costly time wise

File: src/modelfree/train.py [hunk 22]
```diff
@@ -146,14 +148,61 @@ def callback(locals, globals):
...
     # SOMEDAY: If we end up using MPI-based algorithms regularly, come up with a cleaner solution.
     from mpi4py import MPI
     if MPI is None:
         num_proc = 1
     else:
         num_proc = MPI.COMM_WORLD.Get_size()
+    return num_proc
+
+
+class ExpertDatasetOurFormat(ExpertDataset):
```
Comment (decodyng) [inline src/modelfree/train.py]: Another naming nitpick, but from this classname I'd presume that the ExpertDataset itself is actually using our format, instead of what this actually seems to be doing, which is that it's reading in a format in a different style, but the dataset object itself will actually have the same format. Maybe ExpertDatasetFromOurFormat would be better, though I acknowledge that might verge on being too long for a reasonable class name

File: src/modelfree/score_agent.py (hunk 21 above)
Comment (decodyng) [reply src/modelfree/score_agent.py]: [Ignore this, found tests after scrolling down more]

File: tests/modelfree/test_experiments.py [hunk 23]
```diff
@@ -53,6 +59,16 @@ def test_score_agent(config):
     outcomes = [run.result[k] for k in ['ties', 'win0', 'win1']]
     assert sum(outcomes) == run.config['episodes']
 
+    if config.get('record_traj', False):
+        try:
+            for i in range(2):
+                traj_file_path = os.path.join(config['record_traj_params']['save_dir'],
+                                              f'agent_{i}.npz')
+                assert os.path.exists(traj_file_path)
```
Comment (decodyng) [inline tests/modelfree/test_experiments.py]: Don't feel this super strongly, but do somewhat believe that it could be valuable to have this test actually make some assertions about the structure of the numpy object that's being saved, to confirm it matches expectations

File: src/modelfree/common/utils.py [hunk 24]
```diff
@@ -156,6 +171,104 @@ def make_session(graph=None):
     return sess
 
 
+class TrajectoryRecorder(VecMultiWrapper):
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/modelfree/common/utils.py]: Not hard, you can call `env.render(mode='rgb_array')` to get a bitmap image. I'd be concerned this is overlapping with `VideoRecorder` so it might be good to either find a way to line the two up (e.g. with timestamps) or merge the two classes. Feel free to ping me if you want to discuss this further.

File: src/modelfree/common/utils.py [hunk 25]
```diff
@@ -156,6 +171,104 @@ def make_session(graph=None):
...
+            agent_dicts = self.full_traj_dicts[dict_idx]
+            dump_dict = {k: np.asarray(v) for k, v in agent_dicts.items()}
+
+            save_path = os.path.join(save_dir, f'agent_{agent_idx}.npz')
+            np.savez(save_path, **dump_dict)
+
+
 def simulate(venv, policies, render=False):
     """
     Run Environment env with the agents in agents
```
[the remark this answers is not in the document]
Comment (kantneel) [reply src/modelfree/common/utils.py]: Indeed, this is outdated. I think `policies` used to be called `agents` and `venv` was `env`.

File: src/modelfree/common/utils.py [hunk 26]
```diff
@@ -156,6 +171,104 @@ def make_session(graph=None):
...
+        self.record_traj(self.prev_obs, self.actions, rewards, dones, infos)
+        self.prev_obs = observations
+        return observations, rewards, dones, infos
+
+    def reset(self):
+        observations = self.venv.reset()
+        self.prev_obs = observations
+        return observations
+
+    def record_traj(self, prev_obs, actions, rewards, dones, infos):
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/modelfree/common/utils.py]: agreed

[the remark this answers is not in the document]
Comment (kantneel) [reply src/modelfree/common/utils.py]: I like `record_timestep_data`

File: src/modelfree/common/utils.py [hunk 27]
```diff
@@ -156,6 +171,104 @@ def make_session(graph=None):
...
+
+        :param prev_obs: (np.ndarray<float>) observations from previous timestep
+        :param actions: (np.ndarray<float>) actions taken after observing prev_obs
+        :param rewards: (np.ndarray<float>) rewards from actions
+        :param dones: ([bool]) whether episode ended (not recorded)
+        :param infos: ([dict]) dicts with network activations if networks are transparent
+        :return: None
+        """
+        data_keys = ('observations', 'actions', 'rewards')
+        data_vals = (prev_obs, actions, rewards)
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/modelfree/common/utils.py]: agreed

File: src/modelfree/common/utils.py [hunk 28]
```diff
@@ -156,6 +171,104 @@ def make_session(graph=None):
...
+        transparency_keys = ('ff', 'hid')  # we already record observations
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/modelfree/common/utils.py]: This is a good suggestion. I've attempted to implement this feature, was this what you had in mind?

Note I've renamed `transparency_keys` to `info_keys` since this seems more descriptive of what `TrajectoryRecorder` is actually doing. We just happen to be using it for transparency, but we could stuff other values in the `info` dict in the future.

File: src/modelfree/common/utils.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/modelfree/common/utils.py]: yep, a parameter got renamed, fixed.

File: src/modelfree/train.py (hunk 22 above)
Comment (AdamGleave) [reply src/modelfree/train.py]: Since it's only used in this one place I'm OK with a long class name, renamed.

File: src/modelfree/train.py [hunk 29]
```diff
@@ -158,6 +158,48 @@ def _get_mpi_num_proc():
     return num_proc
 
 
+class ExpertDatasetOurFormat(ExpertDataset):
+    """GAIL Expert Dataset. Loads in our format, rather than the GAIL default.
+
+    In particular, GAIL expects a dict of flattened arrays, with episodes concatenated together.
+    The episode start is delineated by an `episode_starts` array.
+
+    By contrast, our format consists of a list of NumPy arrays, one for each episode."""
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/modelfree/train.py]: I've added a reference to the base class. The documentation is currently quite sparse but I have a PR on Stable Baselines open anyway, I'll add some more documentation to it.

File: tests/modelfree/test_experiments.py (hunk 23 above)
Comment (AdamGleave) [reply tests/modelfree/test_experiments.py]: I've added some basic additional tests here.

**Comment by AdamGleave** [comment]
@decodyng I've addressed your comments, requesting another review.

**Comment by AdamGleave** [comment]
Also, I just noticed that `travis_wait` has done something weird with our logging. It no longer prints in real time (seems like a known issue) but more concerningly is also truncating the log, see e.g. https://travis-ci.com/HumanCompatibleAI/adversarial-policies/jobs/194397393

I'm not sure why any of our tests should take any longer than 10 minutes anyway, so I've removed this for now. If we continue to experience problems with Travis *and* have a good reason for a test taking >10 minutes, then we can consider an alternative solution.
